horovod/common/operations.cc

```diff
@@ -1100,29 +1101,34 @@ void CheckForStalledTensors(HorovodGlobalState& state) {
                   << " seconds. ";
         std::cerr << "This may indicate that different ranks are trying to "
                      "submit different tensors or that only subset of ranks is "
-                     "submitting tensors, which will cause deadlock. ";
-        std::cerr << "Stalled ops: ";
+                     "submitting tensors, which will cause deadlock. " << std::endl;
+        std::cerr << "Stalled ops: " << std::endl;
         preamble = true;
-      } else {
-        std::cerr << ", ";
       }
-      std::cerr << tensor_name;
-      std::cerr << " [ready ranks:";
+      std::unordered_set<int32_t> ready_ranks;
+      std::vector<int32_t> missing_ranks;
       for (auto msg_iter = messages.begin(); msg_iter != messages.end();
            msg_iter++) {
-        if (msg_iter == messages.begin()) {
+             ready_ranks.insert(msg_iter->request_rank());
+      }
+      for (int32_t rank = 0; rank < state.size; rank++) {
+        if (ready_ranks.find(rank) == ready_ranks.end()) {
+          missing_ranks.push_back(rank);
```

Comment:
You can combine this and following loop into one and get rid of missing ranks vector.